core/src/main/java/io/atomix/core/queue/impl/DefaultWorkQueueService.java

```diff
@@ -0,0 +1,228 @@
+import io.atomix.primitive.service.BackupInput;
+import io.atomix.primitive.service.BackupOutput;
+import io.atomix.primitive.session.PrimitiveSession;
+import io.atomix.primitive.session.SessionId;
+import io.atomix.utils.serializer.KryoNamespace;
+import io.atomix.utils.serializer.Serializer;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * State machine for {@link WorkQueueProxy} resource.
+ */
+public class DefaultWorkQueueService extends AbstractPrimitiveService<WorkQueueClient> implements WorkQueueService {
+
+  private static final Serializer SERIALIZER = Serializer.using(KryoNamespace.builder()
+      .register((KryoNamespace) WorkQueueType.instance().namespace())
+      .register(TaskAssignment.class)
+      .register(new HashMap().keySet().getClass())
+      .register(ArrayDeque.class)
+      .register(SessionId.class)
+      .build());
+
+  private final AtomicLong totalCompleted = new AtomicLong(0);
+
+  private Queue<Task<byte[]>> unassignedTasks = Queues.newArrayDeque();
+  private Map<String, TaskAssignment> assignments = Maps.newHashMap();
+  private Set<SessionId> registeredWorkers = Sets.newLinkedHashSet();
+
+  public DefaultWorkQueueService() {
+    super(WorkQueueType.instance(), WorkQueueClient.class);
+  }
+
+  @Override
+  public Serializer serializer() {
+    return SERIALIZER;
+  }
+
+  @Override
+  public void backup(BackupOutput writer) {
+    writer.writeObject(registeredWorkers);
+    writer.writeObject(assignments);
+    writer.writeObject(unassignedTasks);
+    writer.writeLong(totalCompleted.get());
+  }
+
+  @Override
+  public void restore(BackupInput reader) {
+    registeredWorkers = reader.readObject();
+    assignments = reader.readObject();
+    unassignedTasks = reader.readObject();
+    totalCompleted.set(reader.readLong());
+  }
+
+  @Override
+  public WorkQueueStats stats() {
+    return WorkQueueStats.builder()
+        .withTotalCompleted(totalCompleted.get())
+        .withTotalPending(unassignedTasks.size())
+        .withTotalInProgress(assignments.size())
+        .build();
+  }
+
+  @Override
+  public void clear() {
+    unassignedTasks.clear();
+    assignments.clear();
+    registeredWorkers.clear();
+    totalCompleted.set(0);
+  }
+
+  @Override
+  public void register() {
+    registeredWorkers.add(getCurrentSession().sessionId());
+  }
+
+  @Override
+  public void unregister() {
+    registeredWorkers.remove(getCurrentSession().sessionId());
+  }
+
+  @Override
+  public void add(Collection<byte[]> items) {
+    AtomicInteger itemIndex = new AtomicInteger(0);
+    items.forEach(item -> {
+      String taskId = String.format("%d:%d:%d",
```

Comment:
Might be cheaper to create a simple key class and use that instead of concatenating three strings.

Reply:
I've also found `String.format` to be incredible costly. It compiles a regex on every call

Reply:
Unfortunately, I probably can't break the API in this case.